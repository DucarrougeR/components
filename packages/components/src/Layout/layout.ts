import {
  border,
  boxShadow,
  color,
  position,
  layout,
  reset,
  space,
  typography,
  BorderProps,
  BoxShadowProps,
  ColorProps,
  PositionProps,
  LayoutProps,
  SpaceProps,
  TypographyProps,
} from 'looker-design-tokens'
import { css } from 'styled-components'

export interface LayoutComponentProps
  extends BorderProps,
    BoxShadowProps,
    ColorProps,
    LayoutProps,
    PositionProps,
    SpaceProps,
    TypographyProps {}

export const layoutCSS = css`
  /**
   * Rules here should provide convenience styling for Box derived components.
   * Generally anything here could be overwritten by explicit values set via
   * Box's prop values. For example a function here that sets 'cursor: pointer'
   * would be overwritten by an explicit <Box cursor='copy'/>.
   */
  ${reset}

  /**
   * Style Utilities that extend Box's props. Most of these come from
   * styled-system but some are Lens specific.
   *
   * These should be last to override rules with lower priority.
   */
  ${border}
  ${boxShadow}
  ${color}
  ${layout}
  ${position}
  ${space}
  ${typography}
`
