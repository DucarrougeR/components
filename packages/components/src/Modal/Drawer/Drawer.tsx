import React from 'react'
import { Modal, ModalProps } from '../Modal'
import { DrawerSurface } from './DrawerSurface'

export const Drawer: React.FC<ModalProps> = ({
  width,
  children,
  surfaceStyles,
  ...props
}) => {
  const surface = (animationState: string) => {
    return (
      <DrawerSurface
        style={surfaceStyles}
        className={animationState}
        width={width}
      >
        {children}
      </DrawerSurface>
    )
  }
  return <Modal {...props} render={surface} />
}
