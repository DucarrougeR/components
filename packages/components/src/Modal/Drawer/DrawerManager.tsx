import React from 'react'
import { ManagedModalProps } from '../Modal'
import { ModalManager } from '../ModalManager'
import { Drawer } from '.'

export class DrawerManager extends ModalManager {
  protected renderModal(content: React.ReactNode, props: ManagedModalProps) {
    return (
      <Drawer isOpen={this.state.isOpen} onClose={this.close} {...props}>
        {content}
      </Drawer>
    )
  }
}
