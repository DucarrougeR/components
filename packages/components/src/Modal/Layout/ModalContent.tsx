import {
  palette,
  CompatibleHTMLProps,
  SpaceProps,
  reset,
  space,
  LayoutProps,
  layout,
} from '@looker/design-tokens'
import React from 'react'
import ReactResizeDetector from 'react-resize-detector'
import styled from 'styled-components'
import omit from 'lodash/omit'

export interface ModalContentProps
  extends LayoutProps,
    CompatibleHTMLProps<HTMLDivElement> {
  /**
   * Style the scrollable space within the ModalContent.
   * Often p="none" is applied if componennts within the the ModalContent need to the
   * touch the container edges.
   */
  innerProps?: SpaceProps
}

interface ContentState {
  overflow: boolean
}

interface InternalContentProps extends ModalContentProps {
  renderedHeight: string
}

class Internal extends React.Component<InternalContentProps, ContentState> {
  private ref: React.RefObject<HTMLDivElement>

  constructor(props: InternalContentProps) {
    super(props)
    this.state = { overflow: false }
    this.ref = React.createRef()
  }

  public hasOverflow(e: HTMLDivElement) {
    return e.offsetHeight < e.scrollHeight
  }

  public componentDidUpdate(prevProps: InternalContentProps) {
    if (prevProps.renderedHeight !== this.props.renderedHeight) {
      this.ref.current &&
        this.setState({ overflow: this.hasOverflow(this.ref.current) })
    }
  }

  public render() {
    const { children, className, innerProps, ...props } = this.props

    if (innerProps && innerProps.p && !innerProps.px) {
      innerProps.px = innerProps.p
    }

    return (
      <Outer
        className={`${className} ${this.state.overflow && 'overflow'}`}
        ref={this.ref}
        {...omit(props, ['renderedHeight'])}
      >
        <Inner {...innerProps}>{children}</Inner>
      </Outer>
    )
  }
}

export const ModalContent = (props: ModalContentProps) => {
  return (
    <ReactResizeDetector handleHeight>
      {(height: string) => <Internal renderedHeight={height} {...props} />}
    </ReactResizeDetector>
  )
}

const Outer = styled.div`
  ${reset}
  ${layout}

  overflow: auto;
  flex: 8;

  &.overflow {
    box-shadow: inset 0 -16px 16px -16px ${palette.charcoal200},
      inset 0 16px 16px -16px ${palette.charcoal200};
  }
`

const Inner = styled.div<SpaceProps>`
  ${reset}
  ${space}
`

Inner.defaultProps = { p: 'large', px: 'xlarge' }
