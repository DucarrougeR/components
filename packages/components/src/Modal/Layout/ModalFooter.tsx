import React, { FC } from 'react'
import styled from 'styled-components'
import { CompatibleHTMLProps } from 'looker-design-tokens'
import { SpaceProps, space } from 'styled-system'

export interface ModalFooterProps extends CompatibleHTMLProps<HTMLDivElement> {
  /**
   * Secondary content to place in the footer
   */
  secondary?: React.ReactNode
}

export const ModalFooter: FC<ModalFooterProps> = ({
  children,
  secondary,
  ...props
}) => {
  return (
    <Layout {...props}>
      <SpaceChildren>{children}</SpaceChildren>
      {secondary && <SpaceChildren mr="auto">{secondary}</SpaceChildren>}
    </Layout>
  )
}

const Layout = styled.footer`
  align-items: center;
  display: flex;
  flex-direction: row-reverse;
  padding: ${props => props.theme.space.large}
    calc(${p => p.theme.space.xlarge} - ${p => p.theme.space.xsmall});
  /**
    * Subtract margin placed on children from the horizontal padding applied to the component
    * so that the content is properly aligned.
    * padding is: vertical: large, horizontal: calc(xlarge - xsmall)
    **/
`

const SpaceChildren = styled.div<SpaceProps>`
  ${space};
  display: flex;
  flex-direction: row-reverse;

  & > * {
    margin-right: ${props => props.theme.space.xsmall};
    margin-left: ${props => props.theme.space.xsmall};
  }
`
