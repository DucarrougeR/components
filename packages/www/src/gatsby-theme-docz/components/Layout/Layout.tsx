import React, { useState } from 'react'
import { Header } from '../Header'
import { Sidebar } from '../Sidebar'
import {
  PageLayout,
  LayoutSidebar,
  LayoutMain,
  ContentArea,
  TopBar,
} from './Layout.styles'

export const Layout: React.FC = ({ children }) => {
  const [open] = useState(true)

  return (
    <PageLayout>
      <TopBar>
        <Header />
      </TopBar>
      <LayoutSidebar>{open && <Sidebar />}</LayoutSidebar>
      <ContentArea>
        <LayoutMain>{children}</LayoutMain>
      </ContentArea>
    </PageLayout>
  )
}
