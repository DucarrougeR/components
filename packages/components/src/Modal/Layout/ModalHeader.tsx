import {
  SpaceProps,
  space,
  CompatibleHTMLProps,
  reset,
} from '@looker/design-tokens'
import { IconNames } from '@looker/icons'
import styled from 'styled-components'
import React, { FC, useContext } from 'react'

import { Box } from '../../Layout/Box'
import { Button } from '../../Button'
import { Heading } from '../../Text'
import { Icon } from '../../Icon'
import { ModalContext } from '../ModalContext'

export interface ModalHeaderProps
  extends SpaceProps,
    CompatibleHTMLProps<HTMLElement> {
  /**
   * Specify an icon to be used for close. Defaults to `Close`
   */
  closeIcon?: IconNames
}

export const ModalHeader: FC<ModalHeaderProps> = ({
  children,
  closeIcon = 'Close',
  ...props
}) => {
  const { closeModal } = useContext(ModalContext)

  return (
    <Header {...props}>
      <Box mr="xlarge">{children}</Box>
      <Button
        tabIndex={-1}
        ml="auto"
        p="none"
        variant="transparent"
        color="neutral"
        onClick={closeModal}
      >
        <Icon name={closeIcon} size="1.25rem" />
      </Button>
    </Header>
  )
}

const Header = styled.header<SpaceProps>`
  ${reset}
  ${space}

  display: flex;
  align-items: center;
  flex-direction: row;

  ${Heading} {
    font-size: ${props => props.theme.fontSizes.medium};
  }
`

Header.defaultProps = {
  p: 'large',
  px: 'xlarge',
}
