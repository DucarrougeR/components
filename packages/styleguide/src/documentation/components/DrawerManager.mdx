DrawerManager gives an easy way to compose a Drawer without the need to manage state.

```js
import { Button } from '../../Button'
import { DrawerManager } from './DrawerManager'
import { ModalContent } from '../'

const content = (
  <ModalContent>
    <p>Some content</p>
    <Button>A button too!</Button>
  </ModalContent>
)
;<DrawerManager content={content}>
  {onClick => <Button onClick={onClick}>Open Drawer</Button>}
</DrawerManager>
```
