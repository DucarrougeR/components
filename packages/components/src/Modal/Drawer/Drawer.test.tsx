import 'jest-styled-components'
import React from 'react'
import {
  assertSnapshot,
  assertSnapshotShallow,
} from 'looker-components-test-utils'
import { ModalContent, ModalFooter, ModalHeader } from '../Layout'
import { Drawer } from './Drawer'
import { DrawerManager } from './DrawerManager'

const content = (
  <>
    words and stuff <button>boo!</button>
  </>
)

test('Drawer Hidden', () => {
  assertSnapshot(
    <DrawerManager content={content}>
      {onClick => <a onClick={onClick}>ðŸ¥‘</a>}
    </DrawerManager>
  )
})

test('Drawer Shown', () => {
  assertSnapshotShallow(<Drawer isOpen>{content}</Drawer>)
})

test('Drawer, backdrop customized', () => {
  assertSnapshotShallow(
    <Drawer isOpen backdrop={{ background: 'purple', opacity: 1 }}>
      {content}
    </Drawer>
  )
})

test('Selection Drawer, Shown', () => {
  assertSnapshotShallow(
    <Drawer isOpen>
      <ModalHeader>Pick stuff</ModalHeader>
      <ModalContent>Stuff and things...</ModalContent>
      <ModalFooter>
        <button>Pick some stuff</button>
        <button>Cancel</button>
      </ModalFooter>
    </Drawer>
  )
})
