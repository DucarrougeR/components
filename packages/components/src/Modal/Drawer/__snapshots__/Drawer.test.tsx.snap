// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer Hidden 1`] = `
<a
  onClick={[Function]}
>
  ðŸ¥‘
</a>
`;

exports[`Drawer Shown 1`] = `ShallowWrapper {}`;

exports[`Drawer, backdrop customized 1`] = `ShallowWrapper {}`;

exports[`Selection Drawer, Shown 1`] = `ShallowWrapper {}`;
